src/linux/component.rs

```diff
@@ -1,160 +1,369 @@
 use crate::ComponentExt;
 
 use std::collections::HashMap;
+use std::convert::TryFrom;
 use std::fs::{metadata, read_dir, File};
 use std::io::Read;
 use std::path::{Path, PathBuf};
 
 #[doc = include_str!("../../md_doc/component.md")]
 pub struct Component {
-    temperature: f32,
-    max: f32,
-    critical: Option<f32>,
+    /// associated device of a Component
+    device: String,
+    /// The chip name.
+    /// 	This should be a short, lowercase string, not containing
+    /// 	whitespace, dashes, or the wildcard character '*'.
+    /// 	This attribute represents the chip name. It is the only
+    /// 	mandatory attribute.
+    /// 	I2C devices get this attribute created automatically.
+    /// 	RO
+    name: String,
+    /// temp[1-*]_input Temperature input value.
+    /// 	Unit: millidegree Celsius
+    /// 	RO
+    temperature: Option<f32>,
+    /// Maximum value computed by sysinfo
+    max: Option<f32>,
+    /// Max threshold provided by the chip/kernel previous user write.
+    /// temp[1-*]_max Temperature min value.
+    ///     Unit: millidegree Celsius
+    ///     RW
+    threshold_max: Option<f32>,
+    /// Max threshold provided by the chip/kernel previous user write.
+    /// temp[1-*]_min Temperature min value.
+    ///     Unit: millidegree Celsius
+    ///     RW
+    threshold_min: Option<f32>,
+    /// temp[1-*]_crit	Temperature critical max value, typically greater than
+    /// 	corresponding temp_max values.
+    /// 	Unit: millidegree Celsius
+    /// 	RW
+    threshold_critical: Option<f32>,
+    /// Rare to have that value
+    /// temp[1-*]_type	Sensor type selection.
+    ///	Integers 1 to 6
+    ///	RW
+    ///	1: CPU embedded diode
+    ///	2: 3904 transistor
+    ///	3: thermal diode
+    ///	4: thermistor
+    ///	5: AMD AMDSI
+    ///	6: Intel PECI
+    ///	Not all types are supported by all chips
+    sensor_type: Option<TermalSensorType>,
+    /// May be empty!
+    /// temp[1-*]_label	Suggested temperature channel label.
+    /// Text string
+    /// Should only be created if the driver has hints about what
+    /// this temperature channel is being used for, and user-space
+    /// doesn't. In all other cases, the label is provided by
+    /// user-space.
+    /// RO
     label: String,
+    // TODO:
+    // temp[1-*]_lowest
+    // Historical minimum temperature
+    // Unit: millidegree Celsius
+    // RO
+    //
+    // temp[1-*]_highest
+    // Historical maximum temperature
+    // Unit: millidegree Celsius
+    //
+    // temp[1-*]_lcrit	Temperature critical min value, typically lower than
+    // corresponding temp_min values.
+    // Unit: millidegree Celsius
+    // RW
+    //
+    // temp[1-*]_emergency
+    // 	Temperature emergency max value, for chips supporting more than
+    // 	two upper temperature limits. Must be equal or greater than
+    // 	corresponding temp_crit values.
+    // 	Unit: millidegree Celsius
+    // 	RW
     input_file: PathBuf,
+    highest_file: PathBuf,
```

Comment:
Instead of keeping empty `PathBuf`, wouldn't it be better to instead have `Option<PathBuf>`?

Reply:
Don't know, always ask myself if I prefers empty `String`, `Vec` etc or `Option`. Would say as you prefers!

Reply:
I am just sure that `Option<f32> is better than NAN x)

Reply:
Have to return to work, will do it tonight or tomorrow morning!

Reply:
had fun with Optionalisation of everything, I am master of `and_then`, `or_else` and try.